server/app/services/pagination/BasePaginationSpec.java

```diff
@@ -0,0 +1,87 @@
+package services.pagination;
+
+import io.ebean.Query;
+
+/**
+ * Abstracts different kind of pagination strategies.
+ *
+ * <p>PaginationSpecs come in 2 major flavors: random acess paging and sequential paging.
+ *
+ * <p>Random Access Paging: requires no knowledge of the last row that was displayed in a previous
+ * page. This spec requires reading from storage all the rows prior to the page being returned,
```

Comment:
(question) What do you mean by "requires reading from storage all the rows"? Does our code not leverage `limit` and `offset` in postgres?

Reply:
lol - -i had written a very nice long reply .. and then I hit "resolve conversation" rather than "add comment".  *sigh*

In order to do an offset, the database has to count the number of rows up to the offset.  An Index can't help here because you don't know what the value of the indexed columns will be at an offset.

The amount of work done to count storage rows up to the offset grows linearly as the offset count increases (increasing page #'s).

So limit does help not read beyond a certain point, but offset only helps not SERVE rows we don't care about, but on the db side you still had to do "work" that grows linearly.

Reply:
Gotcha, the way it's worded sounded like it was all being loaded into memory in the application.